File: scrapers/daycycle.py
import requests
from bs4 import BeautifulSoup
from datetime import date, timedelta
import re
import os
from dotenv import load_dotenv

load_dotenv()

def get_formatted_date(delta_days):
    # Format date as "Tuesday, August 25" (no leading zero)
    d = date.today() + timedelta(days=delta_days)
    return d.strftime('%A, %B %d').replace(' 0', ' ')  # Remove leading zero

def extract_schedule(event_string, current_day):
    schedule_match = re.search(r'HS - ([ABCD]) Schedule', event_string)
    if 'HOLIDAY' in event_string or 'School Closed' in event_string:
        return 'N/A', current_day
    if '1st Day of School' in event_string and current_day is None:
        current_day = 1  # Assume first day of school is Day 1
    if schedule_match and current_day is not None:
        return f"{current_day}{schedule_match.group(1)}", current_day
    return 'N/A', current_day

url = "https://phs.parklandsd.org/about/calendar"
print(f"Fetching data from: {url} at {date.today().isoformat()}")
page = requests.get(url)
if page.status_code != 200:
    print(f"Failed to fetch page: Status {page.status_code}, {page.text}")
    exit(1)

soup = BeautifulSoup(page.content, 'html.parser')
today = get_formatted_date(0)
tomorrow = get_formatted_date(1)
day_after_tomorrow = get_formatted_date(2)
target_dates = [today, tomorrow, day_after_tomorrow]
results = {date: 'N/A' for date in target_dates}
current_day = None  # Track the current day number
school_days = 0  # Track school days to increment day number

for event in soup.find_all("div", class_="fsStateHasEvents"):
    fullstring = event.text.strip().replace("\n", ' ')
    print(f"Processing event: {fullstring}")
    
    # Check if this is a school day (not a holiday)
    if 'HOLIDAY' not in fullstring and 'School Closed' not in fullstring:
        if '1st Day of School' in fullstring:
            current_day = 1  # Reset to Day 1 for the first school day
        elif current_day is not None:
            school_days += 1
            current_day = (school_days % 4) + 1  # Cycle through Day 1-4
    
    # Extract schedule and update current_day
    result, current_day = extract_schedule(fullstring, current_day)
    
    # Check for target dates in the event
    for target_date in target_dates:
        if target_date in fullstring and result != 'N/A':
            results[target_date] = result

# Prepare data for API
final1 = results[today]
final2 = results[tomorrow]
final3 = results[day_after_tomorrow]
print(f"\nScraped Results:")
print(f"Today: {final1}")
print(f"Tomorrow: {final2}")
print(f"Day after tomorrow: {final3}")

# Push data to Vercel API
api_url = os.environ.get('DAYCYCLE_UPDATE_API')
api_key = os.environ.get('API_KEY')
if not api_url:
    print("Error: DAYCYCLE_UPDATE_API not set in environment variables")
    exit(1)
if not api_key:
    print("Error: API_KEY not set in environment variables")
    exit(1)

payload = {
    'today': final1,
    'tomorrow': final2,
    'next_day': final3
}
headers = {
    'Content-Type': 'application/json',
    'api-key': api_key
}
print(f"Sending POST to {api_url} with payload: {payload}")
try:
    response = requests.post(api_url, json=payload, headers=headers, timeout=10)
    print(f"API Response Status: {response.status_code}")
    print(f"API Response Text: {response.text}")
    if response.status_code != 200:
        print(f"Error: Update failed with status {response.status_code}, response: {response.text}")
    else:
        print("Day cycle updated successfully")
except requests.RequestException as e:
    print(f"Network or API error: {str(e)}")

# Print local results
print(f"\nLocal Results:")
print(f"Today: {final1}")
print(f"Tomorrow: {final2}")
print(f"Day after tomorrow: {final3}")

File: scrapers/daycycle.js
const axios = require("axios");
const cheerio = require("cheerio");
const dotenv = require("dotenv");
const path = require("path");

dotenv.config({ path: path.join(__dirname, "..", ".env") });

// Returns YYYY-MM-DD for today + deltaDays (local time)
function dateStr(deltaDays) {
  const d = new Date();
  d.setDate(d.getDate() + deltaDays);
  return d.toISOString().split("T")[0];
}

// Returns "April 24" style string (no leading zero)
function shortDate(deltaDays) {
  const d = new Date();
  d.setDate(d.getDate() + deltaDays);
  const month = d.toLocaleDateString("en-US", { month: "long" });
  return `${month} ${d.getDate()}`;
}

// Returns "Friday, April 24" style string (used in fsStateHasEvents text)
function longDate(deltaDays) {
  const d = new Date();
  d.setDate(d.getDate() + deltaDays);
  const weekday = d.toLocaleDateString("en-US", { weekday: "long" });
  const month = d.toLocaleDateString("en-US", { month: "long" });
  return `${weekday} ${month} ${d.getDate()}`;
}

function parseDayCycle(text) {
  const isHoliday = /HOLIDAY|School Closed|Schools Closed/i.test(text);
  if (isHoliday) return "N/A";

  const dayMatch = text.match(/Day\s+([1-6])/i);
  const schedMatch = text.match(/HS - ([ABCD]) Schedule/i);

  if (dayMatch && schedMatch) {
    return dayMatch[1] + schedMatch[1];
  }
  return "N/A";
}

async function scrapeDayCycle() {
  const today = dateStr(0);
  console.log(`Scraping day cycle for ${today}`);

  // Fetch the calendar grid for the current month via FinalSite elements API.
  // Element 16751 is the main calendar grid on the PHS calendar page.
  // Passing ?date= returns the month containing that date.
  const url = `https://phs.parklandsd.org/fs/elements/16751?date=${today}`;
  console.log(`Fetching: ${url}`);

  let page;
  try {
    page = await axios.get(url, { timeout: 15000 });
  } catch (err) {
    console.error(`Failed to fetch calendar: ${err.message}`);
    process.exit(1);
  }

  const $ = cheerio.load(page.data);

  // Build a map of "Month Day" -> full text for each event day
  const dayMap = {};
  $(".fsStateHasEvents").each((_, el) => {
    const text = $(el).text().trim().replace(/\s+/g, " ");
    // Each element starts with the full date like "Friday April 24 ..."
    // Extract "April 24" portion
    const match = text.match(/(?:Monday|Tuesday|Wednesday|Thursday|Friday|Saturday|Sunday),?\s+([A-Za-z]+ \d+)/);
    if (match) {
      dayMap[match[1]] = text;
    }
  });

  console.log("Days found in calendar:", Object.keys(dayMap).join(", "));

  const results = {
    today: parseDayCycle(dayMap[shortDate(0)] || ""),
    tomorrow: parseDayCycle(dayMap[shortDate(1)] || ""),
    next_day: parseDayCycle(dayMap[shortDate(2)] || ""),
  };

  // If today/tomorrow/next_day spans a month boundary, fetch next month too
  const needsNextMonth = results.tomorrow === "N/A" || results.next_day === "N/A";
  if (needsNextMonth) {
    const nextMonthDate = dateStr(5); // 5 days ahead guaranteed to be in next month if we're near end
    const url2 = `https://phs.parklandsd.org/fs/elements/16751?date=${nextMonthDate}`;
    console.log(`Also fetching next month: ${url2}`);
    try {
      const page2 = await axios.get(url2, { timeout: 15000 });
      const $2 = cheerio.load(page2.data);
      $2(".fsStateHasEvents").each((_, el) => {
        const text = $2(el).text().trim().replace(/\s+/g, " ");
        const match = text.match(/(?:Monday|Tuesday|Wednesday|Thursday|Friday|Saturday|Sunday),?\s+([A-Za-z]+ \d+)/);
        if (match && !dayMap[match[1]]) {
          dayMap[match[1]] = text;
        }
      });
      // Re-compute missing values
      if (results.tomorrow === "N/A") results.tomorrow = parseDayCycle(dayMap[shortDate(1)] || "");
      if (results.next_day === "N/A") results.next_day = parseDayCycle(dayMap[shortDate(2)] || "");
    } catch (err) {
      console.error(`Failed to fetch next month: ${err.message}`);
    }
  }

  console.log("\nScraped Results:");
  console.log(`Today (${shortDate(0)}):     ${results.today}`);
  console.log(`Tomorrow (${shortDate(1)}):  ${results.tomorrow}`);
  console.log(`Next day (${shortDate(2)}):  ${results.next_day}`);

  // Push to API
  const apiUrl = process.env.DAYCYCLE_UPDATE_API;
  const apiKey = process.env.API_KEY;

  if (!apiUrl) {
    console.error("Error: DAYCYCLE_UPDATE_API not set");
    process.exit(1);
  }
  if (!apiKey) {
    console.error("Error: API_KEY not set");
    process.exit(1);
  }

  console.log(`\nSending POST to ${apiUrl} with payload:`, results);
  try {
    const response = await axios.post(apiUrl, results, {
      headers: { "Content-Type": "application/json", "api-key": apiKey },
      timeout: 10000,
    });
    console.log(`API Response Status: ${response.status}`);
    if (response.status === 200) {
      console.log("Day cycle updated successfully");
    } else {
      console.error(`Update failed with status ${response.status}:`, response.data);
    }
  } catch (err) {
    console.error(`Network or API error: ${err.message}`);
  }
}

scrapeDayCycle();
